drivers/rtc/rtc_ll_stm32.c

```diff
@@ -0,0 +1,229 @@
+	|| defined(CONFIG_SOC_SERIES_STM32G0X)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_19
+#elif defined(CONFIG_SOC_SERIES_STM32F4X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F2X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F3X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F7X) \
+	|| defined(CONFIG_SOC_SERIES_STM32WBX) \
+	|| defined(CONFIG_SOC_SERIES_STM32G4X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L1X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H7X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32WLX)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_17
+#endif
+
+/* clang-format on */
+
+#if defined(CONFIG_SOC_SERIES_STM32F1X)
+#error RTC not supported for STM32F1X
+#endif
+
+struct rtc_stm32_config {
+	LL_RTC_InitTypeDef ll_rtc_config;
+	const struct stm32_pclken *pclken;
+};
+
+struct rtc_stm32_data {
+	/* Currently empty */
+};
+
+static int rtc_stm32_init(const struct device *dev)
+{
+	const struct device *const clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	const struct rtc_stm32_config *cfg = dev->config;
+
+	if (!device_is_ready(clk)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	/* Enable RTC bus clock */
+	if (clock_control_on(clk, (clock_control_subsys_t)&cfg->pclken[0]) != 0) {
+		LOG_ERR("clock op failed\n");
+		return -EIO;
+	}
+
+	/* Enable Backup access */
+	z_stm32_hsem_lock(CFG_HW_RCC_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+	LL_PWR_EnableBkUpAccess();
+#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPR_DBP */
+
+	/* Enable RTC clock source */
+	if (clock_control_configure(clk, (clock_control_subsys_t)&cfg->pclken[1], NULL) != 0) {
+		LOG_ERR("clock configure failed\n");
+		return -EIO;
+	}
+
+	LL_RCC_EnableRTC();
+
+	z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
+
+	if (LL_RTC_Init(RTC, ((LL_RTC_InitTypeDef *)&cfg->ll_rtc_config)) != SUCCESS) {
+		return -EIO;
+	}
+
+#ifdef RTC_CR_BYPSHAD
+	LL_RTC_DisableWriteProtection(RTC);
+	LL_RTC_EnableShadowRegBypass(RTC);
+	LL_RTC_EnableWriteProtection(RTC);
+#endif /* RTC_CR_BYPSHAD */
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(CONFIG_CPU_CORTEX_M4)
+	LL_C2_EXTI_EnableIT_0_31(RTC_EXTI_LINE);
+	LL_EXTI_EnableRisingTrig_0_31(RTC_EXTI_LINE);
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* in STM32U5 family RTC is not connected to EXTI */
+#else
+	LL_EXTI_EnableIT_0_31(RTC_EXTI_LINE);
+	LL_EXTI_EnableRisingTrig_0_31(RTC_EXTI_LINE);
+#endif
+
+	return 0;
+}
+
+#if DT_INST_NUM_CLOCKS(0) == 1
+#warning STM32 RTC needs a kernel source clock. Please define it in dts file
+static const struct stm32_pclken rtc_clk[] = {STM32_CLOCK_INFO(0, DT_DRV_INST(0)),
+/* Use Kconfig to configure source clocks fields (Deprecated) */
+/* Fortunately, values are consistent across enabled series */
+#ifdef CONFIG_COUNTER_RTC_STM32_CLOCK_LSE
```

Comment:
This config was for the RTC used as a counter. It has been kept for compatibility reasons but the new preferred way is to define a source clock in dts.
Since this is a new driver, I think we should just support the new way and display an error if a source clock is not provided.

Reply:
Removed.